File: tests/utils/functionalities.ts
// Mirror of `common::functionalities` (programs/common/src/functionalities.rs).
// Flat `u16` identifiers for every instruction in the workspace, excluding
// `factory`. One continuous counter — do not reorder or remove; only append
// new ones at the end, matching the on-chain source.
export const BOND_UPDATE_BOND_TERMS = 0;
export const COUPON_CREATE_COUPON = 1;
export const COUPON_SET_COUPON_RATE = 2;
export const DEACTIVATE_DEACTIVATE = 3;
export const FREEZE_FREEZE_ACCOUNT = 4;
export const FREEZE_UNFREEZE_ACCOUNT = 5;
export const FREEZE_PARTIALLY_FREEZE_ACCOUNT = 6;
export const FREEZE_REMOVE_PARTIAL_FREEZE = 7;
export const METADATA_UPDATE_UPDATE_METADATA_FIELD = 8;
export const METADATA_UPDATE_REMOVE_METADATA_FIELD = 9;
export const MINT_MINT = 10;
export const OPERATIONS_BURN = 11;
export const PAUSE_PAUSE = 12;
export const PAUSE_UNPAUSE = 13;
export const TRANSFER_CONTROL_INITIALIZE = 14;
export const TRANSFER_CONTROL_ADD_TO_WHITELIST = 15;
export const TRANSFER_CONTROL_REMOVE_FROM_WHITELIST = 16;
export const TRANSFER_HOOK_EXECUTE = 17;
export const TREASURY_SET_PAYMENT_TOKEN = 18;
export const TREASURY_PAY_COUPON = 19;
export const ACCESS_CONTROL_GRANT_ROLES = 20;
export const ACCESS_CONTROL_REVOKE_ROLES = 21;

File: programs/common/src/functionalities.rs
/// Flat `u16` identifiers for every public instruction in the workspace, excluding `factory`
/// Values are not scoped per program — so do not reorder or remove an existing constant;
/// only append new ones at the end.
///
/// Named `<PROGRAM>_<INSTRUCTION>`.
pub const BOND_UPDATE_BOND_TERMS: u16 = 0;
pub const COUPON_CREATE_COUPON: u16 = 1;
pub const COUPON_SET_COUPON_RATE: u16 = 2;
pub const DEACTIVATE_DEACTIVATE: u16 = 3;
pub const FREEZE_FREEZE_ACCOUNT: u16 = 4;
pub const FREEZE_UNFREEZE_ACCOUNT: u16 = 5;
pub const FREEZE_PARTIALLY_FREEZE_ACCOUNT: u16 = 6;
pub const FREEZE_REMOVE_PARTIAL_FREEZE: u16 = 7;
pub const METADATA_UPDATE_UPDATE_METADATA_FIELD: u16 = 8;
pub const METADATA_UPDATE_REMOVE_METADATA_FIELD: u16 = 9;
pub const MINT_MINT: u16 = 10;
pub const OPERATIONS_BURN: u16 = 11;
pub const PAUSE_PAUSE: u16 = 12;
pub const PAUSE_UNPAUSE: u16 = 13;
pub const TRANSFER_CONTROL_INITIALIZE: u16 = 14;
pub const TRANSFER_CONTROL_ADD_TO_WHITELIST: u16 = 15;
pub const TRANSFER_CONTROL_REMOVE_FROM_WHITELIST: u16 = 16;
pub const TRANSFER_HOOK_EXECUTE: u16 = 17;
pub const TREASURY_SET_PAYMENT_TOKEN: u16 = 18;
pub const TREASURY_PAY_COUPON: u16 = 19;
pub const ACCESS_CONTROL_GRANT_ROLES: u16 = 20;
pub const ACCESS_CONTROL_REVOKE_ROLES: u16 = 21;

#[cfg(test)]
mod tests {
    /// Parses this file's own source (not a separately maintained list, so it
    /// can't drift from reality) and asserts every `pub const ...: u16 = N;`
    /// value equals its 0-based declaration position. Catches gaps,
    /// duplicates, and out-of-order values — the only valid way to add a
    /// functionality is to append a new constant at the end with the next
    /// number. Shared logic lives in `test_support`.
    #[test]
    fn functionality_constants_are_sequential_from_zero() {
        crate::test_support::assert_u16_constants_sequential_from_zero(
            include_str!("functionalities.rs"),
            "functionality",
        );
    }
}
